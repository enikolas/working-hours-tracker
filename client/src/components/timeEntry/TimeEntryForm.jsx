import React from 'react';
import PropTypes from 'prop-types';

import Button from '../ui/Button';
import InputTimeGroup from '../ui/InputTimeGroup';
import ModeSelect from './ModeSelect';

import strings from '../../utils/strings';
import { Entries } from '../../PropTypes';

const HEADERS = {
  startTime: { key: 0, referenceHour: 8 },
  startBreakTime: { key: 1, referenceHour: 12 },
  endBreakTime: { key: 2, referenceHour: 13 },
  endTime: { key: 3, referenceHour: 17 }
};

const TimeEntryForm = ({
  entry,
  mode,
  isDisabled,
  onChangeEntry,
  onChangeMode,
  isPersisted,
  onSubmit
}) => (
  <form className="TimeEntryForm" onSubmit={onSubmit}>
    <ModeSelect
      mode={mode}
      onSelect={onChangeMode}
    />
    {
      Object.keys(HEADERS).map(phase => (
        <InputTimeGroup
          key={phase}
          isHidden={Boolean(mode)}
          label={strings.times[HEADERS[phase].key].label}
          value={entry[phase]}
          isDisabled={isDisabled}
          onChangeTime={time => onChangeEntry({ ...entry, [phase]: time })}
          referenceHour={HEADERS[phase].referenceHour}
        />
      ))
    }
    <Button
      label={isPersisted ? strings.update : strings.send}
    />
  </form>
);

export default TimeEntryForm;

TimeEntryForm.propTypes = {
  entry: Entries,
  mode: PropTypes.string,
  isDisabled: PropTypes.bool,
  isPersisted: PropTypes.bool,
  onChangeEntry: PropTypes.func,
  onChangeMode: PropTypes.func,
  onSubmit: PropTypes.func
};

TimeEntryForm.defaultProps = {
  entry: {},
  mode: '',
  isDisabled: false,
  isPersisted: false,
  onChangeEntry: () => {},
  onChangeMode: () => {},
  onSubmit: () => {}
};
