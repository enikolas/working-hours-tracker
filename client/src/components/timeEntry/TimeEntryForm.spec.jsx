import React from 'react';
import { shallow } from 'enzyme';

import TimeEntryForm from './TimeEntryForm';

const defaultProps = {
	entry: {
		date: '2018-02-11',
		startTime: '8:00',
		startBreakTime: '11:30',
		endBreakTime: '12:30',
		endTime: '17:00',
		total: '8:00',
		contractedTime: '8:00',
		weekBalance: '20:00',
		balance: '10:00',
		phase: '',
		activity: ''
	},
	isDisabled: false,
	successMessage: '',
	errorMessage: '',
	onChangeEntry: jest.fn(),
	isPersisted: false,
	onSubmit: () => {}
};

describe('TimeEntryForm', () => {
	describe('render', () => {
		it('should render InputTime', () => {
			const wrapper = shallow(<TimeEntryForm
				{...defaultProps}
			/>);
			expect(wrapper).toMatchSnapshot();
		});
		it('should call onChangeEntry', () => {
			const wrapper = shallow(<TimeEntryForm
				{...defaultProps}
			/>);
			const inputTime = wrapper.find('InputTimeGroup').first();
			const onChangeTime = inputTime.prop('onChangeTime');
			const startTime = 8;
			onChangeTime(startTime);
			expect(defaultProps.onChangeEntry).toHaveBeenCalledWith({ ...defaultProps.entry, startTime });
		});
	});
});
