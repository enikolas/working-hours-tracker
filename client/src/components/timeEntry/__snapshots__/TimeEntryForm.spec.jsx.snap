// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeEntryForm render should render InputTime 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TimeEntryForm
    entry={
      Object {
        "activity": "",
        "balance": "10:00",
        "contractedTime": "8:00",
        "date": "2018-02-11",
        "endBreakTime": "12:30",
        "endTime": "17:00",
        "phase": "",
        "startBreakTime": "11:30",
        "startTime": "8:00",
        "total": "8:00",
        "weekBalance": "20:00",
      }
    }
    errorMessage=""
    isDisabled={false}
    isPersisted={false}
    mode=""
    onChangeEntry={[MockFunction]}
    onChangeMode={[Function]}
    onSubmit={[Function]}
    successMessage=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <ModeSelect
          mode=""
          onSelect={[Function]}
        />,
        Array [
          <InputTimeGroup
            isDisabled={false}
            isHidden={false}
            label="Entrada"
            onChangeTime={[Function]}
            referenceHour={8}
            value="8:00"
          />,
          <InputTimeGroup
            isDisabled={false}
            isHidden={false}
            label="Ida para almoço"
            onChangeTime={[Function]}
            referenceHour={12}
            value="11:30"
          />,
          <InputTimeGroup
            isDisabled={false}
            isHidden={false}
            label="Volta do almoço"
            onChangeTime={[Function]}
            referenceHour={13}
            value="12:30"
          />,
          <InputTimeGroup
            isDisabled={false}
            isHidden={false}
            label="Saída"
            onChangeTime={[Function]}
            referenceHour={17}
            value="17:00"
          />,
        ],
        <Button
          isDisabled={false}
          isHidden={false}
          label="Enviar"
        />,
      ],
      "className": "TimeEntryForm",
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "mode": "",
          "onSelect": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "startTime",
        "nodeType": "function",
        "props": Object {
          "isDisabled": false,
          "isHidden": false,
          "label": "Entrada",
          "onChangeTime": [Function],
          "referenceHour": 8,
          "value": "8:00",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "startBreakTime",
        "nodeType": "function",
        "props": Object {
          "isDisabled": false,
          "isHidden": false,
          "label": "Ida para almoço",
          "onChangeTime": [Function],
          "referenceHour": 12,
          "value": "11:30",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "endBreakTime",
        "nodeType": "function",
        "props": Object {
          "isDisabled": false,
          "isHidden": false,
          "label": "Volta do almoço",
          "onChangeTime": [Function],
          "referenceHour": 13,
          "value": "12:30",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "endTime",
        "nodeType": "function",
        "props": Object {
          "isDisabled": false,
          "isHidden": false,
          "label": "Saída",
          "onChangeTime": [Function],
          "referenceHour": 17,
          "value": "17:00",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "isDisabled": false,
          "isHidden": false,
          "label": "Enviar",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <ModeSelect
            mode=""
            onSelect={[Function]}
          />,
          Array [
            <InputTimeGroup
              isDisabled={false}
              isHidden={false}
              label="Entrada"
              onChangeTime={[Function]}
              referenceHour={8}
              value="8:00"
            />,
            <InputTimeGroup
              isDisabled={false}
              isHidden={false}
              label="Ida para almoço"
              onChangeTime={[Function]}
              referenceHour={12}
              value="11:30"
            />,
            <InputTimeGroup
              isDisabled={false}
              isHidden={false}
              label="Volta do almoço"
              onChangeTime={[Function]}
              referenceHour={13}
              value="12:30"
            />,
            <InputTimeGroup
              isDisabled={false}
              isHidden={false}
              label="Saída"
              onChangeTime={[Function]}
              referenceHour={17}
              value="17:00"
            />,
          ],
          <Button
            isDisabled={false}
            isHidden={false}
            label="Enviar"
          />,
        ],
        "className": "TimeEntryForm",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "mode": "",
            "onSelect": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "startTime",
          "nodeType": "function",
          "props": Object {
            "isDisabled": false,
            "isHidden": false,
            "label": "Entrada",
            "onChangeTime": [Function],
            "referenceHour": 8,
            "value": "8:00",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "startBreakTime",
          "nodeType": "function",
          "props": Object {
            "isDisabled": false,
            "isHidden": false,
            "label": "Ida para almoço",
            "onChangeTime": [Function],
            "referenceHour": 12,
            "value": "11:30",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "endBreakTime",
          "nodeType": "function",
          "props": Object {
            "isDisabled": false,
            "isHidden": false,
            "label": "Volta do almoço",
            "onChangeTime": [Function],
            "referenceHour": 13,
            "value": "12:30",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "endTime",
          "nodeType": "function",
          "props": Object {
            "isDisabled": false,
            "isHidden": false,
            "label": "Saída",
            "onChangeTime": [Function],
            "referenceHour": 17,
            "value": "17:00",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "isDisabled": false,
            "isHidden": false,
            "label": "Enviar",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
